### ISSUE TITLE
Create a new material that derives from NeighborCoupleable and TwoMaterialPropertyInterface   

### ISSUE THREAD

**Issue by jiangwen84** [idaholab/moose#12066]
## Rationale

We are developing cohesive zone method (CZM) based on interfaceKernel in MOOSE. For CZM, we need to calculate the displacement jump at the interface and store it as a stateful boundary material property. To calculate displacement jumps, the material need to get the values from both element and its neighbor element. Current Material class only derives from  Coupleable and MaterialPropertyInterface, so maybe we need a new `InterfaceMaterial`, that can derive from `NeighborCoupleable` and `TwoMaterialPropertyInterface`. 

## Description
@lindsayad has started some preliminary work to create a new `MaterialBase` class. The `MaterialBase` class separates out the `Coupleable` and `TwoMaterialPropertyInterface`. The new `Material` and `InterfaceMaterial` derives from  `MaterialBase` and `Coupleable` and `TwoMaterialPropertyInterface`, `NeighborCoupleable ` and `TwoMaterialPropertyInterface `, respectively. This seems to be a good starting point and we should have more design discussions for the following changes. 

## Impact

We have to change some basic internal interfaces.

**Comment by jiangwen84** [comment]
@lindsayad and @andrsd 

In FEProblemBase::addMaterial, I added this line:
std::shared_ptr<MaterialBase> material = _factory.create<MaterialBase>(mat_name, name, parameters, tid);

I had some compile issues, 
error: no viable conversion from 'shared_ptr<MaterialBase>' to 'shared_ptr<Material>'
      _all_materials.addObject(material, tid);

Looks like there are some more changes that we need to make. Any suggestions?

**Comment by jiangwen84** [comment]
tag @arovinelli.

**Comment by lindsayad** [comment]
You're calling:
```c++
template <typename T>
void
MooseObjectWarehouseBase<T>::addObject(std::shared_ptr<T> object)
```
with `T = Material`. We may need to add an additional warehouse, but for the immediate present you could try changing `class MaterialWarehouse : public MooseObjectWarehouse<Material>` to `class MaterialWarehouse : public MooseObjectWarehouse<MaterialBase>`

**Comment by arovinelli** [comment]
@jiangwen84 I don't have a lot of experience about coding at this level in MOOSE so I would try following Alex 's suggestion for the time being. I need to look a bit more into how you are trying to implement this to be more helpful.

Also, if you have other independent tasks required to allow CZM I can work on any of them, as I have some free time now.

**Comment by jiangwen84** [comment]
@arovinelli Sure, I will work with Alex on this. Could you rework on your old PR "Implement an InterfaceUserObject #11660"? It was closed due to inactivity.

**Comment by arovinelli** [comment]
@jiangwen84 I was trying to get back on it right now :)

**Comment by andrsd** [comment]
@jiangwen84 Just for future reference, you really want to add any code into triple back-ticks. It makes a big difference, see:

Without it, we would see
```
error: no viable conversion from 'shared_ptr' to 'shared_ptr'
      _all_materials.addObject(material, tid);
```
with it:
```
error: no viable conversion from 'shared_ptr<MaterialBase>' to 'shared_ptr<Material>'
      _all_materials.addObject(material, tid);
```

Notice how the template argument was eaten (because HTML).

And yes, you want to change the type in `MooseObjectWarehouse` to `MaterialBase`...

**Comment by jiangwen84** [comment]
@andrsd  @lindsayad  Thanks for your suggestions. I changed `std::shared_ptr<Material>` to  `std::shared_ptr<MaterialBase>` in several (maybe all) places. There is one template function in `MooseObjectWarehouseBase` that I am not sure how to change. 

```
template <typename T>
void
MooseObjectWarehouseBase<T>::updateMatPropDependencyHelper(
    std::set<unsigned int> & needed_mat_props, const std::vector<std::shared_ptr<T>> & objects)
{
  for (auto & object : objects)
  {
    auto & mp_deps = object->getMatPropDependencies();
    needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
}
```

The `getMatPropDependencies()` function is implemented in `MaterialPropertyInerface` and `MaterialBase` does not derive from. The compiler error is 

```
/Users/jianw/projects/moose/framework/build/header_symlinks/MooseObjectWarehouseBase.h:648:30: error: no member named 'getMatPropDependencies' in 'MaterialBase'
    auto & mp_deps = object->getMatPropDependencies();
                     ~~~~~~  ^
/Users/jianw/projects/moose/framework/build/header_symlinks/MooseObjectWarehouseBase.h:604:5: note: in instantiation of member function 'MooseObjectWarehouseBase<MaterialBase>::updateMatPropDependencyHelper' requested here
    updateMatPropDependencyHelper(needed_mat_props, getActiveBlockObjects(id, tid));
    ^
/Users/jianw/projects/moose/framework/src/problems/FEProblemBase.C:2367:20: note: in instantiation of member function 'MooseObjectWarehouseBase<MaterialBase>::updateBlockMatPropDependency' requested here
    _all_materials.updateBlockMatPropDependency(blk_id, needed_mat_props, tid);
                   ^
1 error generated.
```

Do you have any suggestions? dynamic_cast and template specialization?

**Comment by andrsd** [comment]
Did you do?
```
- class MaterialWarehouse : public MooseObjectWarehouse<Material>
+ class MaterialWarehouse : public MooseObjectWarehouse<MaterialBase>
```

**Comment by jiangwen84** [comment]
Yes, I did that.

**Comment by andrsd** [comment]
Ok. Do you have the branch somewhere so I can look at it? This might be a little bit more complicated to figure out...

**Comment by jiangwen84** [comment]
I just pushed my branch. Thanks for your time!

**Comment by andrsd** [comment]
This should fix the problem:

[0001-fixup-WIP-Material-MaterialBase-ref-12066.patch.txt](https://github.com/idaholab/moose/files/2354490/0001-fixup-WIP-Material-MaterialBase-ref-12066.patch.txt)

Download the file and do `git am <the file name>` on your branch.

**Comment by jiangwen84** [comment]
@andrsd thanks! I wish I could think of that fix :-)

**Comment by lindsayad** [comment]
@jortensi Use this ticket for communicating any additional needs regarding `InterfaceMaterial`

**Comment by permcody** [comment]
@vincentlaboure - You can see that Alex tagged Javier here. If you have any test cases or other requirements make sure they are clear here.

**Comment by arovinelli** [comment]
@permcody and @lindsayad 
I do have one requirement, this should work also when using a breakmeshbyblocks

**Comment by lindsayad** [comment]
@permcody Is this really supposed to be for FY19 or is this supposed to be for the FY20 milestone? Now is as good a time as any for me to work on this before something else comes up, but I haven't heard from any Rattlesnake folks about what they want this for. E.g. do they want to be able to use this with DGKernels or only InterfaceKernels? Adding @YaqiWang; I would assume he's involved with this...

**Comment by YaqiWang** [comment]
Although our immediate needs are to have interface properties on internal side sets, interface material will be needed by DGKernels as well. Is supporting DGKernels making the implementation harder?

**Comment by lindsayad** [comment]
Yes

**Comment by permcody** [comment]
It needs to be done now but not before the end of FY19.

**Comment by lindsayad** [comment]
> interface material will be needed by DGKernels as well

At the point that we do this, `InterfaceMaterial` will no longer seem like an appropriate name because this material will be acting on subdomain interiors. We essentially want this new material to be able to pull things from corresponding face and neighbor materials. Conceptually it should inherit from `TwoMaterialPropertyInterface`, `NeighborCoupleable`, `BlockRestrictable`, and `BoundaryRestrictable`.

**Comment by YaqiWang** [comment]
It is hard for me to think of this: do we need DGMaterial and InterfaceMaterial like DGKernel and InterfaceKernel?

**Comment by lindsayad** [comment]
> It is hard for me to think of this: do we need DGMaterial and InterfaceMaterial like DGKernel and InterfaceKernel?

That's what I'm thinking...although `InternalSideMaterial` as opposed to `DGMaterial`. Then `InternalSideMaterial` would be `BlockRestrictable` while `InterfaceMaterial` would be `BoundaryRestrictable`

**Comment by arovinelli** [comment]
I agree with @lindsayad. InterfaceMaterial should be boundary restricted and should work with interface kernels, an internal side material should be block restricted and work with DGkernels.
Both should be able to use a disconnected mesh (think for example of using cohesive modeling between each element)

**Comment by lindsayad** [comment]
> Both should be able to use a disconnected mesh

Yea I don't know about that, at least for the initial capability. I thought you had mesh modifiers that allowed you to do cohesive zone with `InterfaceKernel`, e.g. with a regular connectivity mesh?

**Comment by arovinelli** [comment]
@lindsayad Yes we do have the appropriate mesh modifiers for disconnecting the mesh and using the interface kernel.
What I would like is to have the same kind of flexibility for the internal side material.

**Comment by permcody** [comment]
@vincentlaboure, @jortensi - Is there any specific problem you can setup for Alex? We need some guidance on what you need.

**Comment by lindsayad** [comment]
Based on what @YaqiWang said, it sounds like the immediate need is for `InterfaceMaterial` but that eventually they will also want `InternalSideMaterial`. Let me know if I've misunderstood. I have something else I'm working on now, but hopefully I can work on this in parallel.

**Comment by arovinelli** [comment]
@lindsayad and @permcody I can provide a few example problems by the end of the week, both stateful and non stateful

**Comment by lindsayad** [comment]
Based on the title, I consider this closed by #14124. I will create a new issue for the `InternalSideMaterial`
